### PR TITLE
Don't use whole-archive linking

### LINKED ISSUE #70: Use of static libraries and whole-archive causes duplicated symbols
Discovered this while working on #55.  Several months ago, I set up the build system to use a combination of STATIC libraries, linking with `-Wl,--whole-archive`, and `-Wl,--allow-multiple-definition` so that weak symbols would be resolved correctly.  However, this turns out to cause ld to duplicate some/all symbols linked in this manner.  Code still works, but ends up using significantly more memory than it should since everything in Mbed OS gets included twice in the binary.

Full details can be read on the ld bug I filed: https://sourceware.org/bugzilla/show_bug.cgi?id=29660

We need to retool everything to use OBJECT libraries instead of STATIC ones.

### PR DESCRIPTION
**Pull request by multiplemonomials**
### Summary of changes 

This PR is designed to fix #70, see that issue for details. 

It switches the main Mbed OS library to be an OBJECT lib instead of a STATIC lib, since STATIC libs require the dreaded whole-archive option for weak symbol resolution.

Unfortunately, I could not also convert the optional libraries to OBJECT libs because CMake [does not support circular dependencies between OBJECT libraries in any shape or form](https://gitlab.kitware.com/cmake/cmake/-/issues/17905), so the current library structure would not be able to be maintained.  This means that weak symbol resolution between optional libraries and the main library might be problematic.  I claim it's only an issue if the overriding function is defined in a file where no other functions are being referenced in the build (so the entire object will be excluded), but we'll have to test and see.

Note: This PR is a prerequisite for #55, because the NRF52840 linker script doesn't work without this fix,

#### Impact of changes 
Should be no impact for external users, except seeing less memory usage due to the fix for #70.

### Documentation 

None

----------------------------------------------------------------------------------------------------------------
### Pull request type 

    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results 

    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers 

----------------------------------------------------------------------------------------------------------------

### CODE REVIEW

File: tools/cmake/mbed_create_distro.cmake [hunk 1]
```diff
@@ -1,24 +1,8 @@
 # Copyright (c) 2021 ARM Limited. All rights reserved.
 # SPDX-License-Identifier: Apache-2.0
 
-# This script provides mbed_create_distro(), a function that lets you compile multiple
```
Comment (multiplemonomials) [inline tools/cmake/mbed_create_distro.cmake:4]: Removed these docs which were badly out of date

File: connectivity/drivers/emac/TARGET_STM/stm32xx_emac.cpp [hunk 2]
```diff
@@ -1085,4 +1080,46 @@ void HAL_ETH_MACErrorCallback(ETH_HandleTypeDef *heth)
...
+#ifndef USE_USER_DEFINED_HAL_ETH_IRQ_CALLBACK
+
+#define FLAG_RX                 1
+
+/**
+ * Override Ethernet Rx Transfer completed callback
+ * @param  heth: ETH handle
+ * @retval None
+ */
+void HAL_ETH_RxCpltCallback(ETH_HandleTypeDef *heth)
```
Comment (multiplemonomials) [inline connectivity/drivers/emac/TARGET_STM/stm32xx_emac.cpp:1092]: This is an example of code that has to get modified.  These two functions are both weak symbol overrides, and they were declared in a cpp file containing *only* weak symbol overrides, so the linker was discarding that entire file.  So, they have to get moved to a file containing other functions that are actually pulled in to the binary.

**Comment by multiplemonomials** [comment]
Test run for LPC1768:
```
96% tests passed, 3 tests failed out of 72

Total Test time (real) = 1192.02 sec

The following tests did not run:
	  8 - mbed-hal-crc (Skipped)
	 11 - mbed-hal-lp-ticker (Skipped)
	 14 - mbed-hal-ospi (Skipped)
	 16 - mbed-hal-qspi (Skipped)
	 27 - mbed-hal-trng (Skipped)
	 32 - mbed-hal-verify-arduino-pinmap (Skipped)
	 43 - mbed-platform-crash-reporting (Skipped)
	 53 - mbed-platform-stats-cpu (Skipped)
	 54 - mbed-platform-stats-heap (Skipped)
	 55 - mbed-platform-stats-sys (Skipped)
	 56 - mbed-platform-stats-thread (Skipped)
	 71 - mbed-connectivity-nfc-eeprom (Skipped)

The following tests FAILED:
	  3 - mbed-usb-device-hid (Failed)
	  5 - mbed-usb-device-serial (Failed)
	 65 - mbed-rtos-mutex (Failed)
```

Doesn't look like any new failures.

**Comment by multiplemonomials** [comment]
Also ran tests on STM32F429,nothing but USB is failing now that I've implemented the Ethernet fixes

**Review by JohnK1987 — APPROVED** [review]

**Pull request merged**